<a href="https://colab.research.google.com/github/vidjaydas/Data_Science_Python/blob/main/Feature_Engineering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Univariate Selection using Chi Square Test

In [2]:
# Feature Ectraction with Univariate Statistical Test ( Chi-Squared for classification)

from pandas import read_csv
from numpy import set_printoptions
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import chi2

In [7]:
import pandas as pd

data = pd.read_csv('/content/pima-indians-diabetes.data.csv')
data

6  148  72  35    0  33.6  0.627  50  1
0     1   85  66  29    0  26.6  0.351  31  0
1     8  183  64   0    0  23.3  0.672  32  1
2     1   89  66  23   94  28.1  0.167  21  0
3     0  137  40  35  168  43.1  2.288  33  1
4     5  116  74   0    0  25.6  0.201  30  0
..   ..  ...  ..  ..  ...   ...    ...  .. ..
762  10  101  76  48  180  32.9  0.171  63  0
763   2  122  70  27    0  36.8  0.340  27  0
764   5  121  72  23  112  26.2  0.245  30  0
765   1  126  60   0    0  30.1  0.349  47  1
766   1   93  70  31    0  30.4  0.315  23  0

[767 rows x 9 columns]

In [8]:
#Load data
filename = 'pima-indians-diabetes.data.csv'
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
dataframe = read_csv(filename, names=names)
dataframe

preg  plas  pres  skin  test  mass   pedi  age  class
0       6   148    72    35     0  33.6  0.627   50      1
1       1    85    66    29     0  26.6  0.351   31      0
2       8   183    64     0     0  23.3  0.672   32      1
3       1    89    66    23    94  28.1  0.167   21      0
4       0   137    40    35   168  43.1  2.288   33      1
..    ...   ...   ...   ...   ...   ...    ...  ...    ...
763    10   101    76    48   180  32.9  0.171   63      0
764     2   122    70    27     0  36.8  0.340   27      0
765     5   121    72    23   112  26.2  0.245   30      0
766     1   126    60     0     0  30.1  0.349   47      1
767     1    93    70    31     0  30.4  0.315   23      0

[768 rows x 9 columns]

In [12]:
dataframe.shape

(768, 9)

In [13]:
dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   preg    768 non-null    int64  
 1   plas    768 non-null    int64  
 2   pres    768 non-null    int64  
 3   skin    768 non-null    int64  
 4   test    768 non-null    int64  
 5   mass    768 non-null    float64
 6   pedi    768 non-null    float64
 7   age     768 non-null    int64  
 8   class   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [9]:
array = dataframe.values

In [10]:
array

array([[  6.   , 148.   ,  72.   , ...,   0.627,  50.   ,   1.   ],
       [  1.   ,  85.   ,  66.   , ...,   0.351,  31.   ,   0.   ],
       [  8.   , 183.   ,  64.   , ...,   0.672,  32.   ,   1.   ],
       ...,
       [  5.   , 121.   ,  72.   , ...,   0.245,  30.   ,   0.   ],
       [  1.   , 126.   ,  60.   , ...,   0.349,  47.   ,   1.   ],
       [  1.   ,  93.   ,  70.   , ...,   0.315,  23.   ,   0.   ]])

In [22]:
X = array[:,0:8]
Y = array[:,8]    #Class is a dependent variable

In [23]:
#Feature extraction
test = SelectKBest(score_func=chi2, k=4)    #Select best 4 features
fit = test.fit(X, Y)

In [31]:
#Summarize scores
set_printoptions(precision=3)
print(fit.scores_)
features = fit.transform(X)

#For Regression : f_regression, mutual_info_regression
#For Classification : chi2, f_classif, mutual_info_classif


[ 111.52  1411.887   17.605   53.108 2175.565  127.669    5.393  181.304]


####[ 111.52 / 1411.887  / 17.605 /  53.108 / 2175.565 / 127.669  /  5.393 / 181.304]
#### ['preg'/ 'plas'/ 'pres'/ 'skin'/ 'test'/ 'mass'/ 'pedi'/ 'age']

highest score is 2175 i.e. test, it means test is a very important variance, next is 1141.88 i.e. plas and so on

In [32]:
dataframe.head()

preg  plas  pres  skin  test  mass   pedi  age  class
0     6   148    72    35     0  33.6  0.627   50      1
1     1    85    66    29     0  26.6  0.351   31      0
2     8   183    64     0     0  23.3  0.672   32      1
3     1    89    66    23    94  28.1  0.167   21      0
4     0   137    40    35   168  43.1  2.288   33      1

#Recursive Feature Elimination

In [33]:
from pandas import read_csv
from sklearn.feature_selection import RFE
from sklearn.linear_model import LogisticRegression

# Load data

filename = 'pima-indians-diabetes.data.csv'
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
dataframe = read_csv(filename, names=names)

array = dataframe.values
X = array[:,0:8]
Y = array[:,8]

#Feature extraction
model = LogisticRegression(max_iter=400)
rfe = RFE(model, 3)
fit = rfe.fit(X, Y)

In [34]:
#Number Features:
fit.n_features_

3

In [36]:
#Selected Features:
fit.support_       #True means best features

array([ True, False, False, False, False,  True,  True, False])

In [37]:
#Feature Ranking:
fit.ranking_

array([1, 2, 4, 6, 5, 1, 1, 3])

1 means best features, 2 means 2nd best features and so on

In [38]:
dataframe.head()

preg  plas  pres  skin  test  mass   pedi  age  class
0     6   148    72    35     0  33.6  0.627   50      1
1     1    85    66    29     0  26.6  0.351   31      0
2     8   183    64     0     0  23.3  0.672   32      1
3     1    89    66    23    94  28.1  0.167   21      0
4     0   137    40    35   168  43.1  2.288   33      1

#Feature importance using Decision Tree

In [41]:
# Feature Importance with Extra Trees Classifier
from pandas import read_csv
from sklearn.tree import DecisionTreeClassifier

#Load data
filename = 'pima-indians-diabetes.data.csv'
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
dataframe = read_csv(filename, names=names)

array = dataframe.values
X = array[:,0:8]
Y = array[:,8]

# Feature extraction
model = DecisionTreeClassifier()
model.fit(X, Y)
print(model.feature_importances_)

[0.045 0.325 0.091 0.01  0.034 0.243 0.137 0.116]


Highest imformation gain, the more important feature is

.

We will collab all the 3 methods all of the important features and compair which of the features is important in all the 3 methods, then we will consider those methods as the best method